src/NHibernate/Linq/ExpressionExtensions.cs

```diff
@@ -32,5 +33,30 @@ public static bool IsGroupingKeyOf(this MemberExpression expression,GroupResultO
 	
 			return query.QueryModel.ResultOperators.Contains(groupBy);
 		}
+
+		public static bool IsGroupingElementOf(this QuerySourceReferenceExpression expression, GroupResultOperator groupBy)
+		{
+			if (expression.ReferencedQuerySource == groupBy)
+			{
+				return true;
```

Comment:
Odd... that was the original (and only) check that GroupBySelectClauseRewriter.VisitQuerySourceReferenceExpression was doing before I modified it.
No wonder it never worked.